core/config/preconditioner_config.cpp

```diff
@@ -0,0 +1,295 @@
+        if (auto& obj = config.get("l_solver_type")) {
+            str = obj.get_string();
+        }
+        if (str == "solver::LowerTrs") {
+            return dispatch<
+                gko::LinOpFactory,
+                IluHelper2<solver::LowerTrs, solver::UpperTrs,
+                           ReverseApply::value>::template Configurator>(
+                config, context, updated,
+                make_type_selector(updated.get_value_typestr(),
+                                   value_type_list()),
+                make_type_selector(updated.get_index_typestr(),
+                                   index_type_list()));
+        } else if (str == "solver::Ir") {
+            return dispatch<
+                gko::LinOpFactory,
+                IluHelper1<solver::Ir, solver::Ir,
+                           ReverseApply::value>::template Configurator>(
+                config, context, updated,
+                make_type_selector(updated.get_value_typestr(),
+                                   value_type_list()),
+                make_type_selector(updated.get_index_typestr(),
+                                   index_type_list()));
+        } else if (str == "preconditioner::LowerIsai") {
+            return dispatch<
+                gko::LinOpFactory,
+                IluHelper2<preconditioner::LowerIsai, preconditioner::UpperIsai,
+                           ReverseApply::value>::template Configurator>(
+                config, context, updated,
+                make_type_selector(updated.get_value_typestr(),
+                                   value_type_list()),
+                make_type_selector(updated.get_index_typestr(),
+                                   index_type_list()));
+        } else if (str == "solver::Gmres") {
+            return dispatch<
+                gko::LinOpFactory,
+                IluHelper1<solver::Gmres, solver::Gmres,
+                           ReverseApply::value>::template Configurator>(
+                config, context, updated,
+                make_type_selector(updated.get_value_typestr(),
+                                   value_type_list()),
+                make_type_selector(updated.get_index_typestr(),
+                                   index_type_list()));
+        } else {
+            GKO_INVALID_STATE("does not have valid LSolverType");
+        }
+    };
+    bool reverse_apply = false;
+    if (auto& obj = config.get("reverse_apply")) {
+        reverse_apply = obj.get_boolean();
+    }
+    if (reverse_apply) {
+        return dispatch_solver(std::true_type{});
+    } else {
+        return dispatch_solver(std::false_type{});
+    }
+}
+
+
+template <>
+deferred_factory_parameter<gko::LinOpFactory> parse<LinOpFactoryType::Isai>(
+    const pnode& config, const registry& context, const type_descriptor& td)
+{
+    auto updated = update_type(config, td);
+    if (auto& obj = config.get("isai_type")) {
+        auto str = obj.get_string();
+        if (str == "lower") {
```

Comment:
I think for Isai, you could just pass in three type selectors instead of the ifs here.

Reply:
The same might also be possible for Ilu/Ic, but it would require some changes to dispatch, so it's not a high priority.

Reply:
Well, it's not as easy as I thought for the isai (because of the enum). So I guess this is something that can be fixed later.

Reply:
I have setuped the generalized version but it was complained in the old pr.
In short, you can use integral_constant to map value to the type and then decouple it in the dispatch step.
I prefer that way, but I also agree it is tiny usage.